flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java

```diff
@@ -172,4 +217,110 @@ public static TableSchema toSchema(Schema schema) {
 
     return builder.build();
   }
+
+
+   /**
+   * Convert a {@link Schema} to a {@link Schema}.
+   *
+   * @param schema iceberg schema to convert.
+   * @return Flink Schema.
+   */
+  public static org.apache.flink.table.api.Schema toSchema(Schema schema, Map<String, String> properties) {
+
+    org.apache.flink.table.api.Schema.Builder builder = org.apache.flink.table.api.Schema.newBuilder();
+
+    // get watermark and computed columns
+    Map<String, String> watermarkMap = Maps.newHashMap();
+    Map<String, String> computedColumnsMap = Maps.newHashMap();
+    properties.keySet().stream()
+        .filter(k -> k.startsWith(FLINK_PREFIX) && properties.get(k) != null)
+        .forEach(k -> {
+          final String name = k.substring(k.lastIndexOf('.') + 1);
+          String expr = properties.get(k);
+          if (k.startsWith(WATERMARK_PREFIX)) {
+            watermarkMap.put(name, expr);
+          } else if (k.startsWith(COMPUTED_COLUMNS_PREFIX)) {
+            computedColumnsMap.put(name, expr);
+          }
+        });
+
+    // add physical columns.
+    for (RowType.RowField field : convert(schema).getFields()) {
+      builder.column(field.getName(), TypeConversions.fromLogicalToDataType(field.getType()));
+    }
+
+    // add computed columns.
+    computedColumnsMap.forEach(builder::columnByExpression);
+
+    // add watermarks.
+    watermarkMap.forEach(builder::watermark);
+
+    // add primary key.
+    List<String> primaryKey = getPrimaryKeyFromSchema(schema);
+    if (!primaryKey.isEmpty()) {
+      builder.primaryKey(primaryKey.toArray(new String[0]));
+    }
+
+    return builder.build();
+  }
+
+  /**
+   * Convert a {@link CatalogTable} to a {@link ResolvedSchema}.
+   *
+   * @param table flink unresolved schema to convert.
+   * @return Flink ResolvedSchema.
+   */
+  public static ResolvedSchema convertToResolvedSchema(CatalogTable table) {
+    StreamExecutionEnvironment env =  StreamExecutionEnvironment.getExecutionEnvironment(new Configuration());
+    StreamTableEnvironment streamTableEnvironment = StreamTableEnvironment.create(env);
+    CatalogManager catalogManager = ((TableEnvironmentImpl) streamTableEnvironment).getCatalogManager();
+    SchemaResolver schemaResolver = catalogManager.getSchemaResolver();
+    return table.getUnresolvedSchema().resolve(schemaResolver);
```

Comment:
```suggestion
    Configuration configuration = ExecutionEnvironment.getExecutionEnvironment().getConfiguration();
    TableEnvironment tableEnvironment = TableEnvironment.create(configuration);
    Preconditions.checkArgument(table.getDescription().isPresent(), "Illegal table.");
    return tableEnvironment.from(table.getDescription().get()).getResolvedSchema();
```

Based on your comment(https://github.com/apache/iceberg/pull/4246#issuecomment-1110614217), I took a look at our previous proposal and made some suggestions, but it may not be optimal.
Here, I think using 'TableEnvironment' directly is a better choice. It has better commonality, whether streaming, batch, or otherwise, and hides underlying differences.
Let's see if somebody else has a better solution.

Reply:
This method is indeed better!
But the description is taken from the table's comment:
```
    @Override
    public Optional<String> getDescription() {
        return Optional.of(getComment());
    }
```

The comment does not necessarily exist, and if it exists, it is not necessarily the path of the table, so `tableEnvironment.from` may not be able to get the table...

Reply:
Yes, so if we go this way, well, we might need to add a comment here.
https://github.com/apache/iceberg/blob/1300b8ac758e7644861388ac1e90f9ec9bd7e353/flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java#L610-L614

Reply:
The `tableEnvironment` here is a newly created environment with only default_catalog and default_database, but this table actually exists in the catalog in the flink environment, so I think `tableEnvironment.from` should not find the required table. I also actually tested it and couldn't find it.

Reply:
It's a bit tricky. I'll debug it locally to see if there's a better way.